File: r-app/src/services/modules/tokens.ts
import { request } from "../request";

export interface TokenCount {
  inputTokens: number;
}

export const tokensApi = {
  /** payload 形如 { system?, messages: [{role, content}] }。 */
  count: (payload: unknown) =>
    request<TokenCount>("count_tokens", { request: payload }),
};

File: r-app/src-tauri/src/commands/tokens.rs
use serde::Serialize;
use serde_json::{json, Value};

use crate::error::AppResult;
use crate::modules::tokens::estimate_input_tokens;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenCount {
    pub input_tokens: i64,
}

/// 估算请求输入 token（供 `/v1/messages/count_tokens` 与前端工具使用）。
#[tauri::command]
pub fn count_tokens(request: Value) -> AppResult<TokenCount> {
    let system = request.get("system");
    let messages = request.get("messages").cloned().unwrap_or_else(|| json!([]));
    Ok(TokenCount {
        input_tokens: estimate_input_tokens(system, &messages),
    })
}
